### ISSUE TITLE
RSA key lost on leader election (1.7.0)

### ISSUE THREAD

**Issue by sbihel** [hashicorp/nomad#19340]
### Nomad version
`v1.7.0-rc.1`

### Operating system and Environment details
Ubuntu `22.04`
Consul `v1.17.0`

### Issue
I am deploying a fresh cluster and setting up workload identities for Consul and Vault. Bootstrapped the cluster with Nomad `v1.7.0-beta.2` but now all the nodes are using `v1.7.0-rc.1`.

By default, the root keyring contained a single ed25519 key. I then triggered a full rotation to generate an RSA key, but then encountered a few issues:
1. with multiple keys in the keyring, it appears Nomad picks one at random (?), which results in many `failed to verify id token signature` from Consul;
2. the old key never gets out of the `rekeying` state, even though I don't appear to use variables; and
3. after replacing the servers (hoping to address 2.), the new key is now an ed25519 key even though it kept the same UUID.

### Reproduction steps
Deploy a fresh cluster using the latest Nomad version. Perhaps using `v1.7.0-beta.2` at first caused the server to enter a bad state -- I can try to re-bootstrap the cluster if you think this might be the root cause.

#### Expected Result
1. Root key rotation is applied permanently.
2. Root key on a fresh cluster should probably be RS256 directly.

#### Actual Result
1. Root keyring rotation does not stick.
2. Rekeying never stops.

### Nomad Server logs (if appropriate)
When a new server joins the cluster, the only relevant logs appear to be:
```
[WARN]  nomad.keyring.replicator: failed to fetch key from current leader, trying peers: key=128ba7c1-baa0-3bc6-c20f-833b97a1fbe2 error=<nil>
[ERROR] nomad.keyring.replicator: failed to fetch key from any peer: key=128ba7c1-baa0-3bc6-c20f-833b97a1fbe2 error="rpc error: no such key \"128ba7c1-baa0-3bc6-c20f-833b97a1fbe2\" in keyring"
[ERROR] nomad.keyring.replicator: failed to fetch key from any peer: rpc error: no such key "128ba7c1-baa0-3bc6-c20f-833b97a1fbe2" in keyring: key=128ba7c1-baa0-3bc6-c20f-833b97a1fbe2
```

But the `128ba7c1` key was long removed. (I had previously done a non-`-full` rotation.)

### Nomad Client logs (if appropriate)
Error logs related to the id token verification by Consul:
```
msg=
| failed to setup alloc: pre-run hook "consul" failed: 2 errors occurred:
| \t* failed to derive Consul token for service my-service: Unexpected response code: 500 (rpc error making call: error verifying token: failed to verify id token signature)
| \t* failed to derive Consul token for task my-service: Unexpected response code: 500 (rpc error making call: rpc error making call: error verifying token: failed to verify id token signature)
|
failed=true
```

**Comment by sbihel** [comment]
Interesting, but I checked on all servers that `.well-known/jwks.json` would return the new RSA key. Would that contradict your fix? If it doesn't, that would explain the ratio of failure rate

**Comment by tgross** [comment]
Ah, good catch! But as it turns out `/.well-known/jwks.json` gets turned into the `Keyring.ListPublic` RPC, and that's forwarded to the leader! So the mechanism is actually much simpler, and I can reproduce this exactly without doing upgrades or key rotation:

* Run 3 servers
* `curl nomad.example.com:4646/.well-known/jwks.json` that points to any server, see the RSA key
* Restart the leader so that a leader election happens
* `curl nomad.example.com:4646/.well-known/jwks.json`, see the ed25519 key because the leader doesn't have the RSA key from the old leader

So that's a slightly different story than described in #19350 but the same result and same fix. I've updated the PR description.

**Comment by tgross** [comment]
I've merged https://github.com/hashicorp/nomad/pull/19350 and that'll get shipped in Nomad 1.7.1 as soon as feasible. I've renamed this issue to make it easier for folks to find if they've been caught out by it.

**Comment by tgross** [comment]
I've also opened https://github.com/hashicorp/nomad/issues/19367 and https://github.com/hashicorp/nomad/issues/19368 to follow-up on the other two keyring-related behaviors you've described here. I'll follow-up with those separately.

**Comment by sbihel** [comment]
Brilliant, thank you.

**Comment by tgross** [comment]
We've shipped a patch for this bug in Nomad 1.7.1.

**Comment by kaspergrubbe** [comment]
@tgross Thank you so much for fixing this. I am 99% sure that I was hit by this bug, and ended up rebuilding my whole cluster, and then I find this. 

It's completely my fault for missing this in the release notes. But would it be possible to be a bit more loud regarding these footguns in the changelog in the future? And maybe link to a spelled out guide on how to fix it? (maybe it is out there, but I didn't find it).
